util/auth.go

```diff
@@ -16,30 +16,25 @@ package util
 import (
 	"crypto/sha1"
 	"encoding/hex"
+	"fmt"
 
 	"github.com/juju/errors"
 )
 
 // CalcPassword is the algorithm convert hashed password to auth string.
 // See https://dev.mysql.com/doc/internals/en/secure-password-authentication.html
 // SHA1( password ) XOR SHA1( "20-bytes random data from server" <concat> SHA1( SHA1( password ) ) )
-func CalcPassword(scramble, sha1pwd []byte) []byte {
-	if len(sha1pwd) == 0 {
-		return nil
-	}
-	// scrambleHash = SHA1(scramble + SHA1(sha1pwd))
-	// inner Hash
-	hash := Sha1Hash(sha1pwd)
-	// outer Hash
+func CalcPassword(scramble, hpwd, auth []byte) []byte {
 	crypt := sha1.New()
 	crypt.Write(scramble)
-	crypt.Write(hash)
-	scramble = crypt.Sum(nil)
+	crypt.Write(hpwd)
+	hash := crypt.Sum(nil)
 	// token = scrambleHash XOR stage1Hash
-	for i := range scramble {
-		scramble[i] ^= sha1pwd[i]
```

Comment:
I'm confused with this function and the added auth parameter, could you explain it a bit ?